server/src/main/java/org/opensearch/index/engine/InternalEngine.java

```diff
@@ -1075,6 +1059,44 @@ public IndexResult index(Index index) throws IOException {
         }
     }
 
+    @Override
+    public Engine.IndexResult addIndexOperationToTranslog(Index index) throws IOException {
+        try (Releasable ignored = versionMap.acquireLock(index.uid().bytes())) {
```

Comment:
I believe you need to acquire release lock here (as all other public methods do):

```
try (ReleasableLock ignored = readLock.acquire()) {
 ....
}
```

Also, it probably makes sense to do basic checks like `ensureOpen();` and `assertIncomingSequenceNumber` before adding anything to translog.

```
ensureOpen();
assert assertIncomingSequenceNumber(index.origin(), index.seqNo());
```

Another concern I have is that write operations like `index` and `delete` do update `lastWriteNanos` but this is not done here, is there a reason?

Reply:
I'd actually prefer to leave this method out until a segrep PR needs it. Nothing in this PR, except tests, call this method so it doesn't make sense to add at this point until we have a change that's actually using it. That will also concretely communicate why the method is needed to those not intimately working segrep; along with how to best structure the method.

Reply:
:+1: no objections, certainly